keps/sig-node/20191118-cgroups-v2.md

```diff
@@ -0,0 +1,240 @@
+visibility on all the other cgroups.
+
+### Phase 1: Convert from cgroups v1 settings to v2
+
+We can convert the values passed by the k8s in cgroups v1 from to
+cgroups v2 so Kubernetes users don’t have to change what they specify
+in their manifests.
+
+crun has implemented the conversion as follows:
+
+**Memory controller**
+
+| OCI (x) | cgroup 2 value (y) | conversion  |   comment |
+|---|---|---|---|
+| limit | memory.max | y = x ||
+| swap | memory.swap.max | y = x ||
+| reservation | memory.low | y = x ||
+
+**PIDs controller**
+
+| OCI (x) | cgroup 2 value (y) | conversion  |   comment |
+|---|---|---|---|
+| limit | pids.max | y = x ||
+
+**CPU controller**
+
+| OCI (x) | cgroup 2 value (y) | conversion  |  comment |
+|---|---|---|---|
+| shares | cpu.weight | y = (1 + ((x - 2) * 9999) / 262142) | convert from [2-262144] to [1-10000]|
+| period | cpu.max | y = x| period and quota are written together|
+| quota | cpu.max | y = x| period and quota are written together|
+
+**blkio controller**
+
+| OCI (x) | cgroup 2 value (y) | conversion  |   comment |
+|---|---|---|---|
+| weight | io.bfq.weight | y = (1 + (x - 10) * 9999 / 990) | convert linearly from [10-1000] to [1-10000]|
+| weight_device | io.bfq.weight | y = (1 + (x - 10) * 9999 / 990) | convert linearly from [10-1000] to [1-10000]|
+|rbps|io.max|y=x||
+|wbps|io.max|y=x||
+|riops|io.max|y=x||
+|wiops|io.max|y=x||
+
+**cpuset controller**
+
+| OCI (x) | cgroup 2 value (y) | conversion  |   comment |
+|---|---|---|---|
+| cpus | cpuset.cpus | y = x ||
+| mems | cpuset.mems | y = x ||
+
+**hugetlb controller**
+
+| OCI (x) | cgroup 2 value (y) | conversion  |   comment |
+|---|---|---|---|
+| <PAGE_SIZE>.limit_in_bytes | hugetlb.<PAGE_SIZE>.max | y = x ||
+
+With this approach cAdvisor would have to read back values from
+cgroups v2 files (already done).
+
+Kubelet PR: [https://github.com/kubernetes/kubernetes/pull/85218](https://github.com/kubernetes/kubernetes/pull/85218)
+
+### Phase 2: Use cgroups v2 throughout the stack
+
+This option means that the values are written directly to cgroups v2
+by the runtime. The Kubelet doesn’t do any conversion when setting
+these values over the CRI.  We will need to add a cgroups v2 specific
+LinuxContainerResources to the CRI.
+
+This depends upon the container runtimes like runc and crun to be able
+to write cgroups v2 values directly.
+
+OCI will need support for cgroups v2 and CRI implementations will
+write to the cgroups v2 section of the new OCI runtime config.json.
+
+## Risk and Mitigations
+
+Some cgroups v1 features are not available with cgroups v2:
+
+- _cpuacct.usage_percpu_
+- network stats from cgroup
+
+Some cgroups v1 controllers such as _device_ and _net_cls_,
+_net_prio_ are not available with the new version.  The alternative to
+these controllers is to use eBPF.
+
+## Graduation Criteria
+
+- Alpha: Basic support for running Kubernetes on a cgroups v2 host,
```

Comment:
@giuseppe 
all phases(1 and 2), will it be done in alpha?

Reply:
I think we should not block on phase 2 for Alpha, that is going to take much longer.  I'll clarify that